archive/100-theorems-list/90_stirlings_formula.lean

```diff
@@ -0,0 +1,872 @@
+  begin
+    intro n,
+    have h_reorder : ∀ {a b c d e f : ℝ}, a - 1 / (2 : ℝ) * b - c - (d - 1 / (2 : ℝ) * e - f) =
+      (a - d) - 1 / (2 : ℝ) * (b - e) - (c - f),
+    by {intros, ring_nf},
+    rw [log_stirling_seq_formula, log_stirling_seq_formula, h_reorder],
+    repeat {rw [log_div, factorial_succ]},
+    push_cast,
+    repeat {rw log_mul},
+    rw log_exp,
+    ring_nf,
+    all_goals {norm_cast},
+    all_goals {try {refine mul_ne_zero _ _}, try {exact succ_ne_zero _}},
+    any_goals {exact factorial_ne_zero n},
+    any_goals {exact exp_ne_zero 1},
+  end,
+  have h_sum₀, from power_series_log_succ_div m.succ (succ_ne_zero m),
+  have h_nonzero : (m.succ : ℝ) + 1 / (2 : ℝ) ≠ 0,
+  by {rw cast_succ, field_simp, norm_cast, linarith}, --there has to be a better way...
+  rw has_sum_mul_left_iff h_nonzero at h_sum₀,
+  have h_inner : ∀ (b : ℕ) , (((m.succ : ℝ) + 1 / 2) * (2 * (1 / (2 * (b : ℝ) + 1)) *
+      (1 / (2 * (m.succ : ℝ) + 1)) ^ (2 * b + 1)))
+      = (1 : ℝ) / (2 * (b : ℝ) + 1) * ((1 / (2 * m.succ + 1)) ^ 2) ^ b :=
+  begin
+    intro b,
+    rw [← pow_mul, pow_add],
+    have : 2 * (b : ℝ) + 1     ≠ 0, by {norm_cast, exact succ_ne_zero (2*b)},
+    have : 2 * (m.succ :ℝ) + 1 ≠ 0, by {norm_cast, exact succ_ne_zero (2*m.succ)},
+    field_simp,
+    rw mul_comm _ (2 : ℝ),
+    exact mul_rotate' _ _ _,
+  end,
+  have h_sum₁ : has_sum (λ (b : ℕ),
+    ((1 : ℝ) / (2 * (b : ℝ) + 1) * ((1 / (2 * m.succ + 1)) ^ 2) ^ b))
+    (((m.succ : ℝ) + 1 / 2) * log ((m.succ.succ : ℝ) / (m.succ : ℝ))) :=
+  begin
+    refine has_sum.has_sum_of_sum_eq _ h_sum₀,
+    intros,
+    use u,
+    intros v' h,
+    use v',
+    split,
+    { exact h },
+    { refine sum_congr rfl _, intros k hk, exact h_inner k },
+  end,
+  have h_sum : tendsto
+    (λ (n : ℕ), ∑ (k : ℕ) in range n.succ,
+    (λ (b : ℕ), 1 / (2 * (b : ℝ) + 1) * ((1 / (2 * (m.succ : ℝ) + 1)) ^ 2) ^ b) k)
+    at_top (𝓝 (((m.succ : ℝ) + 1 / 2) * log ((m.succ.succ : ℝ) / (m.succ : ℝ)))) :=
+    (has_sum.tendsto_sum_nat h_sum₁).comp (tendsto_add_at_top_nat 1),
+  have split_zero : ∀ (n : ℕ), ∑ (k : ℕ) in range n.succ,
+    1 / (2 * (k : ℝ) + 1) * ((1 / (2 * (m.succ : ℝ) + 1)) ^ 2) ^ k =
+    (∑ (k : ℕ) in range n,
+    1 / (2 * (k.succ : ℝ) + 1) * ((1 / (2 * (m.succ : ℝ) + 1)) ^ 2) ^ k.succ) + 1 := by
+    { intro n,
+      rw sum_range_succ' (λ (k : ℕ),
+        1 / (2 * (k : ℝ) + 1) * ((1 / (2 * (m.succ : ℝ) + 1)) ^ 2) ^ k) n,
+      dsimp only [], --beta reduce functions
+      rw [pow_zero, mul_one, cast_zero, mul_zero, zero_add, div_one] },
+  replace h_sum := tendsto.congr split_zero h_sum,
+  replace h_sum := tendsto.add_const (-1) h_sum,
+  simp only [add_neg_cancel_right] at h_sum, --beta reduce and rw?
+  rw tactic.ring.add_neg_eq_sub _ (1 : ℝ) at h_sum,
+  rw ← hx at h_sum,
+  exact (summable.has_sum_iff_tendsto_nat $
+    (summable_nat_add_iff 1).mpr $ has_sum.summable h_sum₁).mpr h_sum,
+end
```

Comment:
```suggestion
begin
  change
    has_sum ((λ (b : ℕ), 1 / (2 * (b : ℝ) + 1) * ((1 / (2 * (m.succ : ℝ) + 1)) ^ 2) ^ b) ∘ succ) _,
  rw has_sum_nat_add_iff 1,
  convert (power_series_log_succ_div m.succ (succ_ne_zero m)).mul_left ((m.succ : ℝ) + 1 / (2 : ℝ)),
  { ext k,
    rw [← pow_mul, pow_add],
    have : 2 * (k : ℝ) + 1     ≠ 0, by {norm_cast, exact succ_ne_zero (2*k)},
    have : 2 * (m.succ :ℝ) + 1 ≠ 0, by {norm_cast, exact succ_ne_zero (2*m.succ)},
    field_simp,
    ring },
  { have h_reorder : ∀ {a b c d e f : ℝ}, a - 1 / (2 : ℝ) * b - c - (d - 1 / (2 : ℝ) * e - f) =
      (a - d) - 1 / (2 : ℝ) * (b - e) - (c - f),
    by {intros, ring_nf},
    rw [log_stirling_seq_formula, log_stirling_seq_formula, h_reorder],
    repeat {rw [log_div, factorial_succ]},
    push_cast,
    repeat {rw log_mul},
    rw log_exp,
    ring_nf,
    all_goals {norm_cast},
    all_goals {try {refine mul_ne_zero _ _}, try {exact succ_ne_zero _}},
    any_goals {exact factorial_ne_zero m},
    any_goals {exact exp_ne_zero 1},
    simp },
  { apply_instance }
end
```
Two tricks to this gol:
- finding the right lemma (`has_sum_nat_add_iff`)
- reorganizing to avoid the long sequence of transformations `h_sum₀`, `h_sum₁`, `h_sum` -- instead match to something broadly correct using `convert` and then you just have equality goals, which are easier.